packages/textfield/src/mwc-textfield.ts

```diff
@@ -0,0 +1,293 @@
+  notchOutline(value: boolean): void;
+  activateFocus(): void;
+  setTransformOrigin(value: any): void;
+  autoCompleteFocus(): void;
+  deactivateFocus(): void;
+  getValue(): string;
+  setValue(value: string): void;
+  isValid(): boolean;
+  setValid(value: boolean): void;
+  setUseNativeValidation(value: boolean): void;
+  isDisabled(): boolean;
+  setDisabled(value: boolean): void;
+  setHelperTextContent(value: string): void;
+  setLeadingIconAriaLabel(value: string): void;
+  setLeadingIconContent(value: string): void;
+  setTrailingIconAriaLabel(value: string): void;
+  setTrailingIconContent(value: string): void;
+}
+
+export declare var TextFieldFoundation: {
+  prototype: TextFieldFoundation;
+  new (adapter: Adapter): TextFieldFoundation;
+}
+
+@customElement('mwc-textfield' as any)
+export class TextField extends BaseElement {
+  protected mdcFoundation!: TextFieldFoundation;
+
+  protected readonly mdcFoundationClass: typeof TextFieldFoundation = MDCTextFieldFoundation;
+
+  @query('.mdc-text-field')
+  protected mdcRoot!: HTMLElement;
+
+  @query(INPUT_SELECTOR)
+  protected inputSelector!: HTMLInputElement;
+
+  @query(LABEL_SELECTOR)
+  protected labelSelector!: HTMLElement;
+
+  @query(LINE_RIPPLE_SELECTOR)
+  protected lineRippleSelector!: HTMLElement;
+
+  @query(OUTLINE_SELECTOR)
+  protected outlineSelector!: HTMLElement;
+
+  @queryAll(ICON_SELECTOR)
+  protected iconSelector!: HTMLElement[];
+
+  @property({type: String})
+  value = ''
+
+  @property({type: String})
+  label = ''
+
+  @property({type: String})
+  icon = ''
+
+  @property({type: Boolean})
+  iconTrailing = false
+
+  @property({type: Boolean})
+  box = false
+
+  @property({type: Boolean})
+  outlined = false
+
+  @property({type: Boolean})
+  disabled = false
+
+  @property({type: Boolean})
+  fullWidth = false
+
+  @property({type: Boolean})
+  required = false
+
+  @property({type: String})
+  helperText = ''
+
+  @property({type: String})
+  placeHolder = ''
+
+  @property({type: String})
+  type = ''
+
+  @property()
+  lineRipple_
+
+  @property()
+  helperText_
+
+  @property()
+  leadingIcon_
+
+  @property()
+  trailingIcon_
+
+  @property()
+  label_
+
+  @property()
+  outline_
+
+  render() {
+    const classes = {
+      'mdc-text-field--with-leading-icon': this.icon && !this.iconTrailing,
+      'mdc-text-field--with-trailing-icon': this.icon && this.iconTrailing,
+      'mdc-text-field--box': this.box,
+      'mdc-text-field--outlined': !this.fullWidth && this.outlined,
+      'mdc-text-field--disabled': this.disabled,
+      'mdc-text-field--fullwidth': this.fullWidth,
+    };
+    return html`
+      ${this.renderStyle()}
+      <div class="mdc-text-field mdc-text-field--upgraded ${classMap(classes)}">
+        ${!this.fullWidth && this.icon ? html`<i class="material-icons mdc-text-field__icon" tabindex="0">${this.icon}</i>` : ''}
+        ${this._renderInput(this.value, this.required, this.type, this.placeHolder, this.label)}

```

Comment:
Add possibility to embed another component, eg. a button toggling content viability for a password field. See also suggestion with suffix below.
```suggestion
        <slot name="prefix" slot="prefix"></slot>
        ${this._renderInput(this.value, this.required, this.type, this.placeHolder, this.label)}
```